File: src/provide/telemetry/levels.py
# SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
# SPDX-License-Identifier: Apache-2.0
# SPDX-Comment: Part of provide-telemetry.
#

"""The canonical severity ladder and the single string-to-level converter.

Lives at the package root rather than under ``logger/`` so that ``config`` and
``consent`` can use it without importing the logger package, which imports
``config`` in turn.

Python's drift was two separate tables — ``_LEVEL_NAME_TO_NUMERIC`` in
``logger/core.py`` and ``_FAST_LEVEL_LOOKUP`` in ``logger/processors.py`` —
neither of which knew ``WARN``, plus a third ordering in ``consent.py``.
``config`` rejected ``WARN`` outright while Rust accepted it.

See the ``log_levels`` section of ``spec/behavioral_fixtures.yaml`` for the
cross-language contract.
"""

from __future__ import annotations

import logging
from enum import IntEnum

__all__ = [
    "LogSeverity",
    "level_order",
    "parse_level",
    "to_stdlib_level",
    "try_parse_level",
]

#: Numeric level for TRACE, which the stdlib does not define.
TRACE = 5


class LogSeverity(IntEnum):
    """The canonical ladder. The value is the rank, so members compare directly.

    ``WARNING`` and ``FATAL`` are deliberately absent: they are spellings
    resolved by :func:`try_parse_level`, not members. Admitting an alias as a
    member is how ``Warn`` and ``Warning`` both ended up on the public C#
    Logger surface.
    """

    TRACE = 0
    DEBUG = 1
    INFO = 2
    WARN = 3
    ERROR = 4
    CRITICAL = 5

    @property
    def canonical_name(self) -> str:
        """The canonical uppercase spelling."""
        return _CANONICAL_NAMES[self.value]

    @property
    def stdlib_level(self) -> int:
        """The :mod:`logging` numeric level that carries this severity."""
        return _STDLIB_LEVELS[self.value]


# Indexed by rank, so a mutated index is a wrong answer rather than a KeyError.
_CANONICAL_NAMES: tuple[str, ...] = ("TRACE", "DEBUG", "INFO", "WARN", "ERROR", "CRITICAL")

_STDLIB_LEVELS: tuple[int, ...] = (
    TRACE,
    logging.DEBUG,
    logging.INFO,
    logging.WARNING,
    logging.ERROR,
    logging.CRITICAL,
)

# Every accepted spelling, canonical and alias alike. Keys are uppercase; the
# lookup uppercases its input.
_TABLE: dict[str, LogSeverity] = {
    "TRACE": LogSeverity.TRACE,
    "DEBUG": LogSeverity.DEBUG,
    "INFO": LogSeverity.INFO,
    "WARN": LogSeverity.WARN,
    "ERROR": LogSeverity.ERROR,
    "CRITICAL": LogSeverity.CRITICAL,
    "WARNING": LogSeverity.WARN,
    "FATAL": LogSeverity.CRITICAL,
}


def try_parse_level(text: str | None) -> LogSeverity | None:
    """Resolve a level string, or ``None`` when it is not recognised.

    Surrounding whitespace is trimmed and comparison is case-insensitive.
    """
    if text is None:
        return None
    return _TABLE.get(text.strip().upper())


def parse_level(text: str | None, fallback: LogSeverity = LogSeverity.INFO) -> LogSeverity:
    """Resolve a level string, substituting ``fallback`` when unrecognised.

    The fallback is a parameter rather than a hidden constant so the
    substitution is visible at the call site.
    """
    parsed = try_parse_level(text)
    if parsed is None:
        return fallback
    return parsed


def level_order(text: str | None) -> int:
    """Rank a level string, with unrecognised values ranking INFO."""
    return int(parse_level(text))


def to_stdlib_level(level: LogSeverity | str | int) -> int:
    """Coerce any accepted level form to a :mod:`logging` numeric level.

    A :class:`LogSeverity` resolves through the ladder and a string through the
    shared table. A bare ``int`` is passed through as a stdlib level, which is
    what structlog's own ``log()`` has always taken -- so existing callers that
    pass ``logging.WARNING`` keep working.

    The :class:`LogSeverity` check has to come first: ``LogSeverity`` is an
    :class:`~enum.IntEnum`, so every member is also an ``int``. Its values are
    ranks (0-5), not stdlib levels, and ``LogSeverity.CRITICAL`` would
    otherwise pass through as the stdlib's TRACE.
    """
    if isinstance(level, LogSeverity):
        return level.stdlib_level
    if isinstance(level, str):
        return parse_level(level).stdlib_level
    return level

File: go/internal/levelcore/levels.go
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0

// Package levelcore holds the canonical severity ladder and the single place a
// level string becomes a level.
//
// It is shared because the module had three private converters that had
// drifted: the root package's _parseLevel folded CRITICAL onto ERROR, the
// logger package had its own copy of the same fold, and consent's
// _logLevelOrder ranked CRITICAL above ERROR — so a record could be ranked one
// way for filtering and another way for consent inside a single call.
//
// See the log_levels section of spec/behavioral_fixtures.yaml for the
// cross-language contract.
package levelcore

import (
	"log/slog"
	"strings"
)

// Severity is the canonical ladder. The value is the rank, so severities
// compare directly with < and >.
//
// WARNING and FATAL are deliberately absent: they are spellings resolved by
// TryParse, not members.
type Severity uint8

// The ladder, in order.
const (
	Trace Severity = iota
	Debug
	Info
	Warn
	Error
	Critical
)

// canonicalNames is indexed by rank.
var canonicalNames = [...]string{"TRACE", "DEBUG", "INFO", "WARN", "ERROR", "CRITICAL"}

// table maps every accepted spelling, canonical and alias alike, to its rank.
var table = map[string]Severity{
	"TRACE":    Trace,
	"DEBUG":    Debug,
	"INFO":     Info,
	"WARN":     Warn,
	"ERROR":    Error,
	"CRITICAL": Critical,
	"WARNING":  Warn,
	"FATAL":    Critical,
}

// Name returns the canonical uppercase spelling, as it appears on the record.
func (s Severity) Name() string {
	if int(s) >= len(canonicalNames) {
		return canonicalNames[Info]
	}
	return canonicalNames[s]
}

// Order returns the numeric rank, for threshold comparisons.
func (s Severity) Order() uint8 { return uint8(s) }

// TryParse resolves a level string, reporting whether it was recognised.
// Surrounding whitespace is trimmed and comparison is case-insensitive.
func TryParse(text string) (Severity, bool) {
	s, ok := table[strings.ToUpper(strings.TrimSpace(text))]
	if !ok {
		return Info, false
	}
	return s, true
}

// Parse resolves a level string, substituting fallback when it is not
// recognised. The fallback is a parameter rather than a hidden constant so the
// substitution is visible at the call site.
func Parse(text string, fallback Severity) Severity {
	if s, ok := TryParse(text); ok {
		return s
	}
	return fallback
}

// Order returns the rank of a level string, with unrecognised values ranking
// INFO.
func Order(text string) uint8 { return Parse(text, Info).Order() }

// ToSlog converts a severity to the slog level that carries it.
func ToSlog(s Severity) slog.Level {
	// An if-chain rather than a switch: gremlins mutates a switch's case
	// expressions in a way that reports fully covered code as uncovered and
	// fails --threshold-mcover=100. See go/slo.go for the detail.
	if s <= Trace {
		return SlogTrace
	}
	if s <= Debug {
		return slog.LevelDebug
	}
	if s <= Info {
		return slog.LevelInfo
	}
	if s <= Warn {
		return slog.LevelWarn
	}
	if s <= Error {
		return slog.LevelError
	}
	return SlogCritical
}

// FromSlog converts an slog level to the nearest severity at or below it.
func FromSlog(l slog.Level) Severity {
	if l <= SlogTrace {
		return Trace
	}
	if l <= slog.LevelDebug {
		return Debug
	}
	if l <= slog.LevelInfo {
		return Info
	}
	if l <= slog.LevelWarn {
		return Warn
	}
	if l <= slog.LevelError {
		return Error
	}
	return Critical
}

// SlogName is the canonical spelling of an slog level.
//
// slog.Level.String() renders the two custom rungs as "DEBUG-4" and "ERROR+4",
// which no level table recognises. Anything that hands a level to consent, to
// a threshold check, or onto a record must come through here instead.
func SlogName(l slog.Level) string { return FromSlog(l).Name() }

// ParseSlog resolves a level string straight to an slog level.
func ParseSlog(text string, fallback Severity) slog.Level {
	return ToSlog(Parse(text, fallback))
}
